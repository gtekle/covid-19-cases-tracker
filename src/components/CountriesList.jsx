import React, { useEffect, useState, useRef, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { FaChevronLeft, FaSearch } from 'react-icons/fa';

import WORLD_MAP from '../assets/img/world-map.png';
import getCurrentDate from '../utils/currentDate';

import {
  fetchCovidStats,
  filterCountriesByName,
  filterCountriesByPageNumber,
  clearCountriesPerPage
} from '../store/covidStats';
import Country from './Country';
import CustomDatePicker from './CustomDatePicker';

const CountriesList = () => {
  const [countrName, setCountryName] = useState('');
  const [searchInputStatus, setSearchInputStatus] = useState(false);
  const [pageNumber, setPageNumber] = useState(0);
  const pageSize = 8;
  const currentDate = getCurrentDate();
  const dispatch = useDispatch();
  const observer = useRef();
  const { casesByCountry, totalCases, filteredCountries, countriesPerPage } = useSelector((state) => state);
  let alternatingBackgroundColor = 'default_color';
  
  useEffect(() => {
    if (Object.keys(casesByCountry).length === 0) dispatch(fetchCovidStats({ date: currentDate }));
  }, []);
  
  useEffect(() => {
    if (casesByCountry) {
      dispatch(clearCountriesPerPage());
      dispatch(filterCountriesByName({countryName: countrName, casesByCountry}));
    }
    
    if (filteredCountries) dispatch(filterCountriesByPageNumber({pageNumber, pageSize}));

  }, [casesByCountry, countrName]);
  
  useEffect(() => {
    const totalPages = Math.floor(filteredCountries.length / pageSize);
    if (filteredCountries && pageNumber <= totalPages) {
      dispatch(filterCountriesByPageNumber({pageNumber, pageSize}));
    }
	}, [pageNumber]);
  
  const lastCountryElementRef = useCallback(countryNode => {
    if(observer.current) observer.current.disconnect();
    observer.current = new IntersectionObserver( (entries) => {
      if(entries[0].isIntersecting){
        setPageNumber(prevPageNumber => prevPageNumber + 1);
      }
    });
    
    if(countryNode) observer.current.observe(countryNode);
  }, [countriesPerPage]);
  
  const handleChange = (e) => {
    const countryNameValue = e.target.value.trim();
    dispatch(clearCountriesPerPage());
    setCountryName(countryNameValue);
    setPageNumber(0);
  }
  
  const handleSearch = (e) => {
    e.preventDefault();
    setSearchInputStatus(true);
  }
  
  const handleOnSearchInputBlur = (e) => {
    setCountryName('');
    setSearchInputStatus(false);
  }

  const regex = /[*]/i;
  return (
    <div className="countries_list_container">
      <div className="countries_list_header">
        <div className="countries_list_header_back">
          <FaChevronLeft />
          <span>{totalCases.date}</span>
        </div>
        <span>All Stats</span>
        <div className="pick_date">
          { !searchInputStatus &&
            <CustomDatePicker />
          }
          <form>
            { searchInputStatus &&
              <input 
                type="text"
                name="countryName"
                value={countrName}
                id="countryName"
                onChange={ handleChange }
                onBlur={handleOnSearchInputBlur}
                placeholder="search country..."
              />
            }
            <button type="button" value="" onClick={handleSearch} >
              <FaSearch />
            </button>
          </form>
        </div>
      </div>
      <div className="countries_list_hero">
        <img src={WORLD_MAP} alt="world map" />
        <div className="hero_text">
          <h1>Global</h1>
          <span>{totalCases.today_confirmed && totalCases.today_confirmed.toLocaleString('en-US')}</span>
          <br />
          <span>cases</span>
        </div>
      </div>
      <div className="all_stats_divider">
        <span>all stats - todate</span>
      </div>
      <div className="all_stats_detail">
        <span>
          <b>{totalCases.today_confirmed && totalCases.today_confirmed.toLocaleString('en-US')}</b>
          <br />
          Confirmed
        </span>
        <span>
          <b>{totalCases.today_deaths && totalCases.today_deaths.toLocaleString('en-US')}</b>
          <br />
          Deaths
        </span>
        <span>
          <b>{totalCases.today_recovered && totalCases.today_recovered.toLocaleString('en-US')}</b>
          <br />
          Recovered
        </span>
        <span>
          <b>{totalCases.today_open_cases && totalCases.today_open_cases.toLocaleString('en-US')}</b>
          <br />
          Open Cases
        </span>
      </div>
      <div className="stats_by_country_divider">
        <span>stats by country</span>
      </div>
      <ul className="countries_list">
        {
          countriesPerPage.map(
            (country, idx) => {
              if ((idx + 1) % 2 === 0) {
                if (alternatingBackgroundColor === 'default_color') {
                  alternatingBackgroundColor = 'other_color';
                } else {
                  alternatingBackgroundColor = 'default_color';
                }
              }
              return (
                countriesPerPage.length === idx + 1
                ? (
                    <Link ref={lastCountryElementRef} key={country.replace(regex, '')} data-testid={`${country.id}-testId`} to={`/${country.replace(regex, '')}`} className={alternatingBackgroundColor}>
                      <Country country={casesByCountry[country]} />
                    </Link>
                  ) 
                : (
                    <Link key={country} data-testid={`${country.id}-testId`} to={`/${country}`} className={alternatingBackgroundColor}>
                      <Country country={casesByCountry[country]} />
                    </Link>
                  )
              );
            },
          )
        }
      </ul>
    </div>
  );
};

export default CountriesList;
