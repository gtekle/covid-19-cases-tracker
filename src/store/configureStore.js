import { configureStore } from "@reduxjs/toolkit";

import reducer from './covidStats';


export default function () {
  return configureStore({ reducer });
};