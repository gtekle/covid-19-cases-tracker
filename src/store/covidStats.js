import { createSlice } from "@reduxjs/toolkit";

const slice = createSlice({
  name: "covidStats",
  initialState: {casesByCountry: {}, totalCases: {}},
  reducers: {
    statsFetched: (state, action) => {
      console.log("statsFetched");
    }
  }
});

export const { statsFetched } = slice.actions;
export default slice.reducer;