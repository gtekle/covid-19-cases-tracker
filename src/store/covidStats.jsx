import axios from 'axios';

const BASE_URL = 'https://disease.sh/v3/covid-19';
const FETCH_SUCCESS = 'covidStats/covidStatsFetched';
const FILTER_BY_COUNTRY_NAME = 'covidStats/filteredByCountryName';
const FILTER_BY_PAGE_NUMBER = 'covidStats/filteredByPageNumber';
const CLEAER_COUNTRIES_PER_PAGE = 'covidStats/clearedCountriesPerPage';
const FETCH_FAIL = 'covidStats/covidStatsFetchFailed';

export const fetchCovidStats = () => async (dispatch) => {
  let res;
  try {
    const casesByCountryList = await axios.get(`${BASE_URL}/countries`);
    const globalTotals = await axios.get(`${BASE_URL}/all`);

    res = casesByCountryList.data;
    const casesByCountry = casesByCountryList.data;
    const totalCases = globalTotals.data;
    dispatch({
      type: FETCH_SUCCESS,
      payload: {
        casesByCountry,
        totalCases,
      },
    });
  } catch (error) {
    res = error;
    dispatch({
      type: FETCH_FAIL,
      payload: error,
    });
  }

  return res;
};

export const filterCountriesByName = (payload) => ({
  type: FILTER_BY_COUNTRY_NAME,
  payload,
});

export const filterCountriesByPageNumber = (payload) => ({
  type: FILTER_BY_PAGE_NUMBER,
  payload,
});

export const clearCountriesPerPage = () => ({
  type: CLEAER_COUNTRIES_PER_PAGE,
  payload: [],
})

const initialState = {
  casesByCountry: [],
  filteredCountries: [],
  countriesPerPage: [],
  totalCases: {},
};

const covidStatsReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_SUCCESS:
      return {
        ...state,
        casesByCountry: action.payload.casesByCountry,
        totalCases: action.payload.totalCases,
      };
    case FILTER_BY_COUNTRY_NAME:
      return {
        ...state,
        filteredCountries: [ ...action.payload.casesByCountry.filter(
          (country) => country.country.toLowerCase().startsWith(action.payload.countryName.toLowerCase())
          )
        ],
      };
    case FILTER_BY_PAGE_NUMBER:
      const indexOfFirstElement = action.payload.pageNumber*action.payload.pageSize;
      return {
        ...state,
        countriesPerPage: [
          ...state.countriesPerPage,
          ...state.filteredCountries.slice(
            indexOfFirstElement,
            indexOfFirstElement + action.payload.pageSize
          ),
        ]
      };
    case CLEAER_COUNTRIES_PER_PAGE:
      return {
        ...state,
        countriesPerPage: action.payload
      }
    case FETCH_FAIL:
      return initialState;
    default:
      return state;
  }
};

export default covidStatsReducer;
