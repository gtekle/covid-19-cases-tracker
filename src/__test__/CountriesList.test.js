import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';

import store from '../store/configureStore';
import CountriesList from '../components/CountriesList';

describe('Countries List', () => {
  it('CountriesList component renders correctly', () => {
    const tree = render(
      <Provider store={store}>
        <Router>
          <CountriesList />
        </Router>
      </Provider>,
    );
    expect(tree).toMatchSnapshot();
  });

  it('CountriesList component renders correctly', () => {
    const { getByText } = render(
      <Provider store={store}>
        <Router>
          <CountriesList />
        </Router>
      </Provider>,
    );
    const txt = getByText('Recovered');
    expect(txt).toBeInTheDocument();
  });
});
