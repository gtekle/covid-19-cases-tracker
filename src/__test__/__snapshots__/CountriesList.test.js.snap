// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countries List CountriesList component renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="countries_list_container"
      >
        <div
          class="countries_list_header"
        >
          <div
            class="countries_list_header_back"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
              />
            </svg>
            <span />
          </div>
          <span>
            All Stats
          </span>
          <div
            class="pick_date"
          >
            <form>
              <div
                class="d-flex justify-content-around form-group"
              >
                <div
                  class="react-datepicker-wrapper"
                >
                  <div
                    class="react-datepicker__input-container"
                  >
                    <input
                      class=""
                      name="startDate"
                      type="text"
                      value="2022-03-11"
                    />
                  </div>
                </div>
                <button
                  type="submit"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 16 16"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M6.445 11.688V6.354h-.633A12.6 12.6 0 0 0 4.5 7.16v.695c.375-.257.969-.62 1.258-.777h.012v4.61h.675zm1.188-1.305c.047.64.594 1.406 1.703 1.406 1.258 0 2-1.066 2-2.871 0-1.934-.781-2.668-1.953-2.668-.926 0-1.797.672-1.797 1.809 0 1.16.824 1.77 1.676 1.77.746 0 1.23-.376 1.383-.79h.027c-.004 1.316-.461 2.164-1.305 2.164-.664 0-1.008-.45-1.05-.82h-.684zm2.953-2.317c0 .696-.559 1.18-1.184 1.18-.601 0-1.144-.383-1.144-1.2 0-.823.582-1.21 1.168-1.21.633 0 1.16.398 1.16 1.23z"
                    />
                    <path
                      d="M3.5 0a.5.5 0 0 1 .5.5V1h8V.5a.5.5 0 0 1 1 0V1h1a2 2 0 0 1 2 2v11a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V3a2 2 0 0 1 2-2h1V.5a.5.5 0 0 1 .5-.5zM1 4v10a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V4H1z"
                    />
                  </svg>
                </button>
              </div>
            </form>
          </div>
        </div>
        <div
          class="countries_list_hero"
        >
          <img
            alt="world map"
            src="world-map.png"
          />
          <div
            class="hero_text"
          >
            <h1>
              Global
            </h1>
            <span />
            <br />
            <span>
              cases
            </span>
          </div>
        </div>
        <div
          class="all_stats_divider"
        >
          <span>
            all stats - todate
          </span>
        </div>
        <div
          class="all_stats_detail"
        >
          <span>
            <b />
            <br />
            Confirmed
          </span>
          <span>
            <b />
            <br />
            Deaths
          </span>
          <span>
            <b />
            <br />
            Recovered
          </span>
          <span>
            <b />
            <br />
            Open Cases
          </span>
        </div>
        <div
          class="stats_by_country_divider"
        >
          <span>
            stats by country
          </span>
        </div>
        <ul
          class="countries_list"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="countries_list_container"
    >
      <div
        class="countries_list_header"
      >
        <div
          class="countries_list_header_back"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
            />
          </svg>
          <span />
        </div>
        <span>
          All Stats
        </span>
        <div
          class="pick_date"
        >
          <form>
            <div
              class="d-flex justify-content-around form-group"
            >
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class=""
                    name="startDate"
                    type="text"
                    value="2022-03-11"
                  />
                </div>
              </div>
              <button
                type="submit"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.445 11.688V6.354h-.633A12.6 12.6 0 0 0 4.5 7.16v.695c.375-.257.969-.62 1.258-.777h.012v4.61h.675zm1.188-1.305c.047.64.594 1.406 1.703 1.406 1.258 0 2-1.066 2-2.871 0-1.934-.781-2.668-1.953-2.668-.926 0-1.797.672-1.797 1.809 0 1.16.824 1.77 1.676 1.77.746 0 1.23-.376 1.383-.79h.027c-.004 1.316-.461 2.164-1.305 2.164-.664 0-1.008-.45-1.05-.82h-.684zm2.953-2.317c0 .696-.559 1.18-1.184 1.18-.601 0-1.144-.383-1.144-1.2 0-.823.582-1.21 1.168-1.21.633 0 1.16.398 1.16 1.23z"
                  />
                  <path
                    d="M3.5 0a.5.5 0 0 1 .5.5V1h8V.5a.5.5 0 0 1 1 0V1h1a2 2 0 0 1 2 2v11a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V3a2 2 0 0 1 2-2h1V.5a.5.5 0 0 1 .5-.5zM1 4v10a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V4H1z"
                  />
                </svg>
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="countries_list_hero"
      >
        <img
          alt="world map"
          src="world-map.png"
        />
        <div
          class="hero_text"
        >
          <h1>
            Global
          </h1>
          <span />
          <br />
          <span>
            cases
          </span>
        </div>
      </div>
      <div
        class="all_stats_divider"
      >
        <span>
          all stats - todate
        </span>
      </div>
      <div
        class="all_stats_detail"
      >
        <span>
          <b />
          <br />
          Confirmed
        </span>
        <span>
          <b />
          <br />
          Deaths
        </span>
        <span>
          <b />
          <br />
          Recovered
        </span>
        <span>
          <b />
          <br />
          Open Cases
        </span>
      </div>
      <div
        class="stats_by_country_divider"
      >
        <span>
          stats by country
        </span>
      </div>
      <ul
        class="countries_list"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
